opt_einsum/paths.py

```diff
@@ -193,154 +210,213 @@ def _optimal_iterate(path, remaining, inputs, flops):
+                if new_flops < self.best_progress[len(inputs)]:
+                    self.best_progress[len(inputs)] = new_flops
+                # sieve based on current progress relative to best
+                elif new_flops > self.cutoff_flops_factor * self.best_progress[len(inputs)]:
+                    return None
 
-            # sieve based on current best flops
-            new_flops = flops + flops12
-            if new_flops >= best['flops']:
-                return None
-
-            # compare to how the best method was doing as this point
-            if new_flops < best_progress[len(inputs)]:
-                best_progress[len(inputs)] = new_flops
-            # sieve based on current progress relative to best
-            elif new_flops > 4 * best_progress[len(inputs)]:
-                return None
-
-            # create a sort order based on cost
-            size1, size2 = size_cache[inputs[i]], size_cache[inputs[j]]
-            try:
-                size12 = size_cache[k12]
-            except KeyError:
-                size12 = size_cache[k12] = helpers.compute_size_by_dict(k12, size_dict)
-
-            # sieve based on memory limit
-            if (memory_limit not in _UNLIMITED_MEM) and (size12 > memory_limit):
-                # terminate path here, but check all-terms contract first
-                new_flops = flops + _compute_oversize_flops(inputs, remaining, output, size_dict)
-                if new_flops < best['flops']:
-                    best['flops'] = new_flops
-                    best['path'] = path + (tuple(remaining),)
-                return None
-
-            # set cost heuristic in order to locally sort possible contractions
-            cost = size12 - size1 - size2
-            return cost, flops12, new_flops, (i, j), k12
+                # sieve based on memory limit
+                if (memory_limit not in _UNLIMITED_MEM) and (size12 > memory_limit):
+                    # terminate path here, but check all-terms contract first
+                    new_flops = flops + _compute_oversize_flops(inputs, remaining, output, size_dict)
+                    if new_flops < self.best['flops']:
+                        self.best['flops'] = new_flops
+                        self.best['ssa_path'] = path + (tuple(remaining),)
+                    return None
 
-        # check all possible remaining paths
-        candidates = []
-        for i, j in itertools.combinations(remaining, 2):
-            if i > j:
-                i, j = j, i
-            k1, k2 = inputs[i], inputs[j]
+                # set cost heuristic in order to locally sort possible contractions
+                size1, size2 = size_cache[inputs[i]], size_cache[inputs[j]]
 
-            # initially ignore outer products
-            if k1.isdisjoint(k2):
-                continue
+                if self.cost_fn is None:
+                    cost = size12 - size1 - size2
```

Comment:
Maybe add this cost function to the dictionary set if none in the init? Would simplify the logic here.

Reply:
Yep, I've simplified the logic for the ``cost_fn`` and ``choose_fn`` [now](https://github.com/dgasmith/opt_einsum/pull/78/commits/18df4be675746daaa203c84e59f26bfdc15c92d8), which has also reduced any overhead to negligible.